File: game-server/src/main/java/net/dodian/uber/game/persistence/PlayerSaveRepository.java
package net.dodian.uber.game.persistence;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;

import static net.dodian.utilities.DatabaseKt.getDbConnection;

public class PlayerSaveRepository {

    @FunctionalInterface
    public interface ConnectionProvider {
        Connection getConnection() throws SQLException;
    }

    private final ConnectionProvider connectionProvider;

    public PlayerSaveRepository() {
        this(PlayerSaveRepository::defaultConnection);
    }

    public PlayerSaveRepository(ConnectionProvider connectionProvider) {
        this.connectionProvider = connectionProvider;
    }

    private static Connection defaultConnection() throws SQLException {
        return getDbConnection();
    }

    public void saveSnapshot(PlayerSaveSnapshot snapshot) throws SQLException {
        try (Connection connection = connectionProvider.getConnection()) {
            boolean originalAutoCommit = connection.getAutoCommit();
            connection.setAutoCommit(false);

            try (Statement statement = connection.createStatement()) {
                statement.executeUpdate(snapshot.getStatsUpdateSql());

                if (snapshot.getStatsProgressInsertSql() != null) {
                    statement.executeUpdate(snapshot.getStatsProgressInsertSql());
                }

                statement.executeUpdate(snapshot.getCharacterUpdateSql());
                connection.commit();
            } catch (SQLException sqlException) {
                connection.rollback();
                throw sqlException;
            } catch (RuntimeException runtimeException) {
                connection.rollback();
                throw runtimeException;
            } finally {
                connection.setAutoCommit(originalAutoCommit);
            }
        }
    }
}

File: game-server/src/main/kotlin/net/dodian/uber/game/persistence/player/PlayerSaveRepository.kt
package net.dodian.uber.game.persistence.player

import java.sql.Connection
import java.sql.SQLException
import net.dodian.uber.game.persistence.db.dbConnection

class PlayerSaveRepository(
    private val connectionProvider: ConnectionProvider = ConnectionProvider { defaultConnection() },
) {
    fun interface ConnectionProvider {
        @Throws(SQLException::class)
        fun getConnection(): Connection
    }

    @Throws(SQLException::class)
    fun saveSnapshot(snapshot: PlayerSaveSnapshot) {
        connectionProvider.getConnection().use { connection ->
            val originalAutoCommit = connection.autoCommit
            connection.autoCommit = false
            try {
                connection.createStatement().use { statement ->
                    statement.executeUpdate(snapshot.statsUpdateSql)
                    snapshot.statsProgressInsertSql?.let(statement::executeUpdate)
                    statement.executeUpdate(snapshot.characterUpdateSql)
                    connection.commit()
                }
            } catch (sqlException: SQLException) {
                connection.rollback()
                throw sqlException
            } catch (runtimeException: RuntimeException) {
                connection.rollback()
                throw runtimeException
            } finally {
                connection.autoCommit = originalAutoCommit
            }
        }
    }

    companion object {
        @JvmStatic
        @Throws(SQLException::class)
        fun defaultConnection(): Connection = dbConnection
    }
}
